### PR TITLE
executor: refine merge join

### PR DESCRIPTION
**Pull request by XuHuaiyu**
e.outerIter4Row should not change if e.innerIter4Row not reach the end during join.
PTAL @zz-jason

### CODE REVIEW

**Comment by zz-jason** [comment]
please add a test

**Comment by zz-jason** [comment]
```sql
create table t(a bigint, b bigint);
insert into t values(1, 1), (1, 2), (1, 3);
create index idx_a on t(a);
analyze table t;
set @@tidb_max_chunk_size=2;
select /*+ TIDB_SMJ(t2) */ count(*) from t t1 inner join t t2 on t1.a = t2.a;
```

**Comment by XuHuaiyu** [comment]
PTAL @zz-jason @winoros @coocood

**Comment by zz-jason** [comment]
LGTM

**Comment by XuHuaiyu** [comment]
Before this commit, the result of the test case added in the code would get `3`.

File: executor/merge_join.go [hunk 1]
```diff
@@ -523,10 +523,12 @@ func (e *MergeJoinExec) joinToChunk(chk *chunk.Chunk) (hasMore bool, err error)
 		for ; e.outerIter4Row.Current() != e.outerIter4Row.End(); e.outerIter4Row.Next() {
 			err = e.resultGenerator.emitToChunk(e.outerIter4Row.Current(), e.innerIter4Row, chk)
 			if err != nil || chk.NumRows() == e.maxChunkSize {
-				e.outerIter4Row.Next()
+				if e.innerIter4Row.Current() == e.innerIter4Row.End() {
```
Comment (coocood) [inline executor/merge_join.go]: We can move this if branch out of the outer if, after `emitToChunk`, then remove `e.outerIter4Row.Next` in line 523 and remove line 530.

**Comment by XuHuaiyu** [comment]
PTAL @coocood

**Pull request merged**